registry/storage/catalog.go

```diff
@@ -71,33 +62,69 @@ func (reg *registry) Repositories(ctx context.Context, repos []string, last stri
 
 // Enumerate applies ingester to each repository
 func (reg *registry) Enumerate(ctx context.Context, ingester func(string) error) error {
-	repoNameBuffer := make([]string, 100)
+	repos := make([]string, 100)
 	var last string
-	for {
-		n, err := reg.Repositories(ctx, repoNameBuffer, last)
-		if err != nil && err != io.EOF {
-			return err
-		}
+	var namespacePaths []string
 
-		if n == 0 {
-			break
-		}
+	root, err := pathFor(repositoriesRootPathSpec{})
+	if err != nil {
+		return err
+	}
 
-		last = repoNameBuffer[n-1]
-		for i := 0; i < n; i++ {
-			repoName := repoNameBuffer[i]
-			err = ingester(repoName)
-			if err != nil {
+	// Shallow walk first to build buffer of namespaces
```

Comment:
After studying this closer, `Walk` is already doing this. Saving the child list here won't provide benefit.

The main performance benefit is that we have a single call to walk.